addons/l10n_fr_hr_holidays/models/hr_leave.py

```diff
@@ -0,0 +1,135 @@
+
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from collections import defaultdict
+
+from odoo import fields, models, api
+
+class HrLeave(models.Model):
+    _inherit = 'hr.leave'
+
+    l10n_fr_date_to = fields.Datetime('End Date For French Rules', readonly=True)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            employee = self.env['hr.employee'].browse(vals['employee_id']).sudo()
+            employee_calendar = employee.resource_calendar_id
+            company_calendar = employee.company_id.resource_calendar_id
+            if employee.company_id.country_id.code == 'FR' and employee_calendar != company_calendar:
+                vals['date_to'] = self._get_fr_new_date_to(vals['date_to'], employee_calendar, company_calendar)
+        return super().create(vals_list)
+
+    def _calendar_works_on_date(self, CalendarAttendance, calendar, working_days, date):
+        dayofweek = str(date.weekday())
+        if calendar.two_weeks_calendar:
+            weektype = str(CalendarAttendance.get_week_type(date))
+            return working_days[weektype][dayofweek]
+        return working_days[False][dayofweek]
+
+    def _get_working_hours(self, calendar):
+        working_days = defaultdict(lambda: defaultdict(lambda: False))
+        for attendance in calendar.attendance_ids:
+            working_days[attendance.week_type][attendance.dayofweek] = True
+        return working_days
+
+    def _is_on_part_time_off(self, CalendarAttendance, employee_calendar, company_calendar, date):
+        company_working_days = self._get_working_hours(company_calendar)
+        if not self._calendar_works_on_date(CalendarAttendance, company_calendar, company_working_days, date):
+            return False
+        employee_working_days = self._get_working_hours(employee_calendar)
+        if self._calendar_works_on_date(CalendarAttendance, employee_calendar, employee_working_days, date):
+            return False
+        return True
+
+    def _get_fr_new_date_to(self, date_to, employee_calendar, company_calendar):
+        employee_working_days = self._get_working_hours(employee_calendar)
+        company_working_days = self._get_working_hours(company_calendar)
+
+        CalendarAttendance = self.env['resource.calendar.attendance']
+        date_target = date_to
+        if isinstance(date_to, str):
+            date_target = datetime.fromisoformat(date_to)
+        new_date_to = date_target
+        date_target += relativedelta(days=1)
+        while not self._calendar_works_on_date(CalendarAttendance, employee_calendar, employee_working_days, date_target):
+            if self._calendar_works_on_date(CalendarAttendance, company_calendar, company_working_days, date_target):
+                new_date_to = date_target
+            date_target += relativedelta(days=1)
+
+        return ' '.join(new_date_to.isoformat().split('T'))
```

Comment:
We should get rid of this isoformat thing.

Reply:
It should be fixed!